velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -129,21 +151,50 @@ class FirstAggregate : public FirstLastAggregateBase<numeric, TData> {
       const SelectivityVector& rows,
       const std::vector<VectorPtr>& args,
       bool /* mayPushdown */) override {
-    DecodedVector decoded(*args[0], rows);
+    this->decodedValue_.decode(*args[0], rows);
+
+    rows.applyToSelected([&](vector_size_t i) {
+      updateValue(this->decodedValue_.index(i), groups[i], this->decodedValue_);
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    this->decodeIntermediateRows(rows, args);
 
-    rows.applyToSelected(
-        [&](vector_size_t i) { updateValue(i, groups[i], decoded); });
+    rows.applyToSelected([&](vector_size_t i) {
+      updateValue(
+          this->decodedIntermediates_.index(i), groups[i], this->decodedValue_);
+    });
   }
 
   void addSingleGroupRawInput(
       char* group,
       const SelectivityVector& rows,
       const std::vector<VectorPtr>& args,
       bool /* mayPushdown */) override {
-    DecodedVector decoded(*args[0], rows);
+    this->decodedValue_.decode(*args[0], rows);
 
-    rows.testSelected(
-        [&](vector_size_t i) { return updateValue(i, group, decoded); });
+    rows.testSelected([&](vector_size_t i) {
+      return updateValue(
+          this->decodedValue_.index(i), group, this->decodedValue_);
+    });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    this->decodeIntermediateRows(rows, args);
+
+    rows.testSelected([&](vector_size_t i) {
+      return updateValue(
```

Comment:
I wonder if there is a bug.

updateValue uses 'index' to check for null in this->decodedValue_, but it seems like it should be for null in this->decodedIntermediates_ instead.

Reply:
In addXXXIntermediatesResults, I use decodedIntermediates_ to decode row(T, boolean), and use decodedValue_ to decode T (first column).
I learn this usage from below code and found we can use the index of decodedIntermediates_ to check its child's null.

https://github.com/facebookincubator/velox/blob/e8c33912b5cf9693d49241a5f2c7d9d496bc1470/velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp#L326-L335 

Hope you could give more insights my usage is wrong, thanks!

Reply:
This is correct, but don't we need to check if the row of the RowVector itself is null?

Reply:
I got this problem in early implementation, and I add set null logic for row vector in extractAccumulators before set null for first column vector, see Line#104. So, I think it is safe to only check null for first column vector here.

Reply:
And I add lots of tests to ensure this logic is safe.

Reply:
Got it. Just to clarify, you are certain that RowVector itself is never null. Thanks.